### PR TITLE
feat(form-core): 5-10x faster `makePathArray`

### PR DESCRIPTION
**Pull request by GiacoCorsiglia**
## 🎯 Changes

_NB: I used Claude Code when doing this work (but not to write this description)._

In https://github.com/TanStack/form/issues/2150 I noted that mounting `<form.Field>` is slow because it exhibits O(N^2) complexity, where N is the number of fields in the form.  This PR does **not** fix the O(N^2) behavior, but when profiling it, I noticed that `makePathArray` is a super hot path.

On `main`, `makePathArray` includes a bunch of regex string munging.  I rewrote it as a single-pass for loop.  This produces 5–10x speed up in microbenchmarks, and what appears to be ~2x faster `<form.Field>` mounting/unmounting.

I also wrote additional tests for `makePathArray` to try to capture its edge case behavior (what I think are malformed inputs—things not allowed by `DeepKeys<T>`).  I did my best to maintain backwards compatibility, but you will see there is one BC-break I identified:

```
// Old behavior:
expect(makePathArray('a]b')).toEqual(['ab'])
// New behavior:
expect(makePathArray('a]b')).toEqual(['a', 'b'])
```

I may have missed some edge cases not covered by my new tests.  It's possible to match the old behavior more closely.  Claude's original implementation was much more complex and handled more edge cases, but I opted to simplify so that the code was easier to understand.  Happy to adjust.

### Benchmarks

This branch includes benchmark files I would not expect to be merged, but wanted to include temporarily so others can validate my results (run on my M4 Pro MackBook Pro).

Here's the output for the new `utils.bench.ts`:

```
 BENCH  Summary

   @tanstack/form-core  old - tests/utils.bench.ts > array input (fast path, no parsing)
    1.10x faster than new

   @tanstack/form-core  new - tests/utils.bench.ts > simple key (no nesting)
    7.56x faster than old

   @tanstack/form-core  new - tests/utils.bench.ts > uuid key
    4.78x faster than old

   @tanstack/form-core  new - tests/utils.bench.ts > dot notation
    5.90x faster than old

   @tanstack/form-core  new - tests/utils.bench.ts > mixed dot and bracket notation
    11.81x faster than old

   @tanstack/form-core  new - tests/utils.bench.ts > deeply nested mixed path
    11.61x faster than old

   @tanstack/form-core  new - tests/utils.bench.ts > numeric string with leading zeros (kept as string)
    9.95x faster than old

   @tanstack/form-core  new - tests/utils.bench.ts > numeric string (converted to number)
    9.67x faster than old
```

In addition, I (well, Claude) wrote `field-render-perf.test.tsx`, which is an automated version of the [reproduction](https://stackblitz.com/edit/vitejs-vite-bbfkr8kq?file=src%2FApp.tsx,src%2FApp.css,src%2Findex.css&terminal=dev) I put together for #2150.  

Results with the old `makePathArray()`:

```
iterations=5 warmup=1 env=jsdom (median of 5)
    N      mount    unmount      total        total min..max
  100     17.9ms     18.5ms     36.3ms          35.3..45.8ms
  500    185.1ms    419.9ms    601.5ms        599.8..612.1ms
 1000    691.2ms   1897.7ms   2549.5ms      2457.9..2677.3ms
 2000   2709.0ms   8351.3ms  11060.6ms    10882.0..11462.1ms
```

Results with the new `makePathArray()`:

```
iterations=5 warmup=1 env=jsdom (median of 5)
    N      mount    unmount      total        total min..max
  100     13.2ms      7.6ms     20.9ms          18.1..26.8ms
  500     86.0ms    130.4ms    218.2ms        213.2..232.6ms
 1000    293.7ms    719.7ms   1015.8ms       929.1..1081.8ms
 2000   1190.4ms   3499.1ms   4705.3ms      4456.6..4951.0ms
```

~2x faster!

## ✅ Checklist

- [x] I have followed the steps in the [Contributing guide](https://github.com/TanStack/form/blob/main/CONTRIBUTING.md).
- [x] I have tested this code locally with `pnpm test:pr`.

## 🚀 Release Impact

- [x] This change affects published code, and I have generated a [changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md).
- [ ] ~This change is docs/CI/dev-only (no release).~

## Summary by CodeRabbit

* **Performance**
  * Improved form field mounting and unmounting performance.

[Review Change Stack](https://app.coderabbit.ai/change-stack/TanStack/form/pull/2152?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/react-form/tests/field-render-perf.test.tsx
Comment (GiacoCorsiglia) [inline packages/react-form/tests/field-render-perf.test.tsx:1]: I would expect to delete this file before merging.

File: packages/react-form/tests/field-render-perf.test.tsx [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+ *
+ *   # 2) CPU profile of a single mount + unmount, written next to this file as
+ *   #    field-render-perf.mount.N{N}.cpuprofile / .unmount.N{N}.cpuprofile.
+ *   #    Open with Chrome DevTools (Performance → Load profile) or VS Code's
+ *   #    "vscode-js-profile-flame" extension.
+ *   PERF=1 PROFILE=1 PROFILE_N=2000 pnpm --filter @tanstack/react-form test:lib field-render-perf
+ *
+ * (No `--` separator: with pnpm 10's `--filter`, args after `--` get dropped
+ * before vitest sees them, so the filename filter is silently ignored and the
+ * full suite runs instead.)
```
Comment (GiacoCorsiglia) [inline packages/react-form/tests/field-render-perf.test.tsx]: You can tell Claude wrote this goofy comment 😅 

But you can verify that the benchmark results match the real-world behavior visible [here](https://stackblitz.com/edit/vitejs-vite-bbfkr8kq?file=src%2FApp.tsx,src%2FApp.css,src%2Findex.css&terminal=dev).

File: packages/form-core/vite.config.ts
Comment (GiacoCorsiglia) [inline packages/form-core/vite.config.ts:1]: This can be dropped if we drop the benchmarks

File: packages/form-core/package.json
Comment (GiacoCorsiglia) [inline packages/form-core/package.json:1]: This can be dropped if we drop the benchmarks

File: packages/form-core/tests/utils.bench.ts
Comment (GiacoCorsiglia) [inline packages/form-core/tests/utils.bench.ts:1]: I imagine this will be removed before this PR is merged (should that happen)

File: packages/form-core/tests/utils.bench.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import { bench, describe } from 'vitest'
+import { makePathArray } from '../src/utils'
+
+// Snapshot of the original implementation for side-by-side comparison.
+// Remove this and the paired benches once the new implementation is merged.
+const reLineOfOnlyDigits = /^(\d+)$/gm
```
[the remark this answers is not in the document]
Comment (GiacoCorsiglia) [reply packages/form-core/tests/utils.bench.ts]: Correct.

File: packages/react-form/tests/field-render-perf.test.tsx [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+ * RESULTS BEFORE OPTIMIZATION ==================================
+ *  iterations=5 warmup=1 env=jsdom (median of 5)
+ *      N      mount    unmount      total        total min..max
+ *    100     17.9ms     18.5ms     36.3ms          35.3..45.8ms
+ *    500    185.1ms    419.9ms    601.5ms        599.8..612.1ms
+ *   1000    691.2ms   1897.7ms   2549.5ms      2457.9..2677.3ms
+ *   2000   2709.0ms   8351.3ms  11060.6ms    10882.0..11462.1ms
+ * ==============================================================
+ *
+ * NOTE: This file is intended to be removed before merge.
```
[the remark this answers is not in the document]
Comment (GiacoCorsiglia) [reply packages/react-form/tests/field-render-perf.test.tsx]: I leave it to the maintainers whether they want such a thing!

File: packages/react-form/tests/field-render-perf.test.tsx [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+      for (let i = 0; i < WARMUP; i++) await measureOnce(ids)
+
+      let result: ReturnType<typeof render> | null = null
+      const mountPath = join(
+        import.meta.dirname,
+        `field-render-perf.mount.N${PROFILE_N}.cpuprofile`,
+      )
+      const unmountPath = join(
+        import.meta.dirname,
+        `field-render-perf.unmount.N${PROFILE_N}.cpuprofile`,
```
[the remark this answers is not in the document]
Comment (GiacoCorsiglia) [reply packages/react-form/tests/field-render-perf.test.tsx]: This file is not intended to land in `main`

**Comment by crutchcorn** [comment]
This is awesome, thank you!

Consider the new code "LGTM"'d. Let's go ahead and remove the `old` code, remove benchmarking, and generally clean up the PR to land in `main`

Let me know what you need from us!

File: .gitignore [hunk 5]
```diff
@@ -28,6 +28,7 @@ size-plugin.json
 stats-hydration.json
 stats.json
 stats.html
+*.cpuprofile
```
Comment (GiacoCorsiglia) [inline .gitignore:31]: Figured I might as well leave this

**Comment by GiacoCorsiglia** [comment]
@crutchcorn Amazing!  I removed all the benchmarks and cleaned up the comments in the new tests, so I think it's ready to be merged now.  But let me know if you need anything else.

**Comment by RobertBeekman** [comment]
Just a little bump as it has been 2 weeks. Would be great to see this merged, we've run into the same performance bottleneck :)

**Review by crutchcorn — APPROVED** [review]
Sorry for delay in merging - been overwhelmed lately with other stuff. Once it passes CI, we'll merge and release
